Libraries/LibWeb/DOM/Element.cpp

```diff
@@ -2690,6 +2690,150 @@ bool Element::is_relevant_to_the_user()
     return false;
 }
 
+// https://drafts.csswg.org/css-contain-2/#skips-its-contents
+bool Element::skips_its_contents()
+{
+    // https://drafts.csswg.org/css-contain-2/#valdef-content-visibility-hidden
+    // The element skips its contents.
+    if (computed_properties()->content_visibility() == CSS::ContentVisibility::Hidden)
+        return true;
+
+    // https://drafts.csswg.org/css-contain-2/#valdef-content-visibility-auto
+    // If the element is not relevant to the user, it also skips its contents.
+    if (computed_properties()->content_visibility() == CSS::ContentVisibility::Auto && !this->is_relevant_to_the_user()) {
+        return true;
+    }
+
+    return false;
+}
+
+// https://drafts.csswg.org/css-contain-2/#containment-types
+
+// https://drafts.csswg.org/css-contain-2/#containment-size
+bool Element::has_size_containment() const
+{
+    // However, giving an element size containment has no effect if any of the following are true:
+
+    // - if the element does not generate a principal box (as is the case with 'display: contents' or 'display: none')
+    if (!layout_node())
+        return false;
+
+    // - if its inner display type is 'table'
+    if (layout_node()->display().is_table_inside())
+        return false;
+
+    // - if its principal box is an internal table box
+    if (layout_node()->display().is_internal_table())
+        return false;
+
+    // - if its principal box is an internal ruby box or a non-atomic inline-level box
+    // FIXME: Implement this.
+
+    if (computed_properties()->contain().contains_slow(CSS::Contain::Size))
+        return true;
+
+    return false;
+}
+// https://drafts.csswg.org/css-contain-2/#containment-inline-size
+bool Element::has_inline_size_containment() const
+{
+    // Giving an element inline-size containment has no effect if any of the following are true:
+
+    // - if the element does not generate a principal box (as is the case with 'display: contents' or 'display: none')
+    if (!layout_node())
+        return false;
+
+    // - if its inner display type is 'table'
+    if (layout_node()->display().is_table_inside())
+        return false;
+
+    // - if its principal box is an internal table box
+    if (layout_node()->display().is_internal_table())
+        return false;
+
+    // - if its principal box is an internal ruby box or a non-atomic inline-level box
+    // FIXME: Implement this.
+
+    if (computed_properties()->contain().contains_slow(CSS::Contain::InlineSize))
```

Comment:
I'm not 100% sure on this, but it sounds like `contain: size` implies `contain: inline-size`, so we should check for that as well here. That's unless you want to distinguish between the two somehow.

Reply:
I don't think so.
As the spec is written, size containment and inline-size containment are two distinct things that can be enabled for an element and nowhere does it say that `contain: size` should enable inline-size containment.
They just happen to have some of the same effects. (size containment along the inline axis)

Reply:
Yeah, them having the same side-effects is what I was thinking. An optimisation that works for `inline-size` would also want to be enabled for `size` I think. But we can worry about that once such an optimisation exists.

Reply:
Fair. I'm mainly worried that any other spec, at any point, can decide to care about specifically inline-size containment and not regular size containment.
It seems unlikely, and I don't know why any spec would want to do that, but I also don't feel confident enough to rule it out.